packages/optimizer/lib/transformers/OptimizeImages.js

```diff
@@ -0,0 +1,249 @@
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {hasAttribute, nextNode, firstChildByTag} = require('../NodeUtils');
+const {skipNodeAndChildren} = require('../HtmlDomHelper');
+const {isValidImageSrcURL} = require('../URLUtils');
+
+const MAX_IMG_SIZE = 820;
+const MIN_WIDTH_TO_ADD_SRCSET_IN_RESPONSIVE_LAYOUT = 300;
+const NUM_SRCSET_DPR = [1.0, 2.0, 3.0];
+const SRCSET_WIDTH = [
+  39,
+  47,
+  56,
+  68,
+  82,
+  100,
+  120,
+  150,
+  180,
+  220,
+  270,
+  330,
+  390,
+  470,
+  560,
+  680,
+  820,
+  1000,
+  1200,
+];
+
+/**
+ * Default implementation that does not perform any kind of image optimization, but generates a `srcset` string
+ * by appending the width to the file name (e.g. image.jpg => image.32w.jpg).
+ *
+ * @param {string} src - the image src
+ * @param {number} width - the required widths (in px)
+ * @returns {string|undefined} - the image URL or undefined if no image is available in this dimension
+
+ */
+const DEFAULT_IMAGE_OPTIMIZER = (src, width) => {
+  // we cannot rename if the image does not have a file extension
+  const index = src.lastIndexOf('.');
+  if (index === -1) {
+    return null;
+  }
+  const prefix = src.substring(0, index);
+  const postfix = src.substring(index, src.length);
+  return `${prefix}.${width}w${postfix}`;
+};
+
+/**
+ *
+ */
+class SrcsetWidth {
+  constructor() {
+    this.widthList_ = [];
+  }
+
+  /**
+   * Sets the base width, i.e., renderered dimension measured in CSS pixels.
+   * Returns true if srcset is needed, that is, we'll resize the image to at
+   * least 2 legitimate widths.
+   * if max_img_width is provided the actual image size in srcset will not
+   * exceed this value. So if max_img_width is 820, the srcset will not
+   * contain any image greater than 820px. The max_img_width is not absolute
+   * number but depends on the aspect ratio. So if 650 is max_img_width, the
+   * nearest aspect ratio width for this max width is 620.
+   *
+   * @param {Number} imgSrcWidth
+   * @param {Number} maxImgWidth
+   */
+  setBaseWidth(imgSrcWidth, maxImgWidth = -1) {
+    this.widthList_.length = 0;
+    let previousWidth = -1;
+    if (maxImgWidth > 0 && imgSrcWidth > maxImgWidth) {
+      return;
+    }
+
+    for (let i = NUM_SRCSET_DPR.length - 1; i >= 0; --i) {
+      let width = this.roundUp(Math.ceil(imgSrcWidth * NUM_SRCSET_DPR[i]));
+      if (maxImgWidth > 0 && width > maxImgWidth) {
+        width = maxImgWidth;
+      }
+      if (width != previousWidth) {
+        this.widthList_.push(width);
+      }
+
+      previousWidth = width;
+    }
+  }
+
+  /**
+   *  Returns true if there is more legitimate width.
+   */
+  moreWidth() {
+    return this.widthList_.length > 0;
+  }
+
+  /**
+   * Returns the current legitimate width and moves the state to the next one.
+   */
+  nextWidth() {
+    const nextWidth = this.widthList_[this.widthList_.length - 1];
+    this.widthList_.pop();
+    return nextWidth;
```

Comment:
any reason this wouldn't just be something like
```js
return this.widthList_.pop()
```

or is the variable allocation for readability?

Reply:
no reason...